test/widgets/compose_box_test.dart

```diff
@@ -238,4 +242,134 @@ void main() {
+        final sendButtonFinder = find.ancestor(
+          of: find.byIcon(Icons.send),
+          matching: find.byType(IconButton));
+        final sendButtonWidget = tester.widget<IconButton>(sendButtonFinder);
+        final sendButtonElement = tester.element(sendButtonFinder);
+        final colorScheme = Theme.of(sendButtonElement).colorScheme;
+        final expectedForegroundColor = expected
+          ? colorScheme.onSurface.withOpacity(0.38)
+          : colorScheme.onPrimary;
+        check(sendButtonWidget.color).equals(expectedForegroundColor);
+      }
+
+      testWidgets('success', (tester) async {
+        final controllerKey = await prepareComposeBox(tester, StreamNarrow(eg.stream().streamId));
+        final composeBoxController = controllerKey.currentState!;
+
+        // (When we check that the send button looks disabled, it should be because
+        // the file is uploading, not because the topic is missing.)
+        composeBoxController.topicController!.value = const TextEditingValue(text: 'some topic');
+
+        testBinding.pickFilesResult = FilePickerResult([PlatformFile(
+          readStream: Stream.fromIterable(['asdf'.codeUnits]),
+          // TODO test inference of mimeType from initial bytes, when
+          //   it can't be inferred from path
+          path: '/private/var/mobile/Containers/Data/Application/foo/tmp/image.jpg',
+          name: 'image.jpg',
+          size: 12345,
+        )]);
+        connection.prepare(delay: const Duration(seconds: 1), json:
+          UploadFileResult(uri: '/user_uploads/1/4e/m2A3MSqFnWRLUf9SaPzQ0Up_/image.jpg').toJson());
+
+        await tester.tap(find.byIcon(Icons.image));
+        final call = testBinding.takePickFilesCalls().single;
+        check(call.allowMultiple).equals(true);
+        check(call.type).equals(FileType.media);
+
+        final errorDialogs = tester.widgetList(find.byType(AlertDialog));
+        check(errorDialogs).isEmpty();
+
+        check(composeBoxController.contentController.text).equals('[Uploading image.jpg...]()\n\n');
+        // (the request is checked more thoroughly in API tests)
+        check(connection.lastRequest!).isA<http.MultipartRequest>()
+          ..method.equals('POST')
+          ..files.single.which((it) => it
+            ..field.equals('file')
+            ..length.equals(12345)
+            ..filename.equals('image.jpg')
+            ..contentType.asString.equals('image/jpeg')
+            ..has<Future<List<int>>>((f) => f.finalize().toBytes(), 'contents')
+              .completes((it) => it.deepEquals(['asdf'.codeUnits].expand((l) => l)))
+          );
+        checkAppearsLoading(tester, true);
+
+        await tester.pump(const Duration(seconds: 1));
+        const expectedContent = '[image.jpg](/user_uploads/1/4e/m2A3MSqFnWRLUf9SaPzQ0Up_/image.jpg)\n\n';
+        checkAppearsLoading(tester, false);
+        check(composeBoxController.contentController.text).equals(expectedContent);
```

Comment:
nit: bit simpler if expected gets inlined:
```suggestion
        check(composeBoxController.contentController.text)
          .equals('[image.jpg](/user_uploads/1/4e/m2A3MSqFnWRLUf9SaPzQ0Up_/image.jpg)\n\n');
        checkAppearsLoading(tester, false);
```

plus it's less indented so a bit easier to read.